### DISCUSSION TITLE
Retreiving entities next to one entity

### DISCUSSION THREAD

**Discussion by mc-lep** [Doraku/DefaultEcs#171] [Q&A] [answered]
Hi,
First you did a very good job with this ECS, thanks you.

I've just started with DefaultECS and ECS in general. I'm trying to figure out how to implement a collision system.
The issue I'm facing is the broad phase aka how to query entities next to the entity I'm processing.

I wish to implement a simple spatial grid and store the entities on that grid based on their 2d position (like this one : https://gameprogrammingpatterns.com/spatial-partition.html)

In your documentation you state that I should avoid storing entities, so I have difficulties to figure out how to implement this kind of system.

Have you a way to acheive that ?

**Comment by genaray** [comment]
Not the owner of this repo, but i also use default ECS and a quadtree.
I did it like this... i give EVERY entity an unique id and attach it to them. Its like a hash or GUID basically... and i map every entity to that id for O(1) performance when searching the entity by its unique id.

Then i basically grapped the first best quadtree library and put thos entities by their unique id in it. Then you can query nearby entities, receive their unique ids, resolve them to get the real entity references.

**Comment by Doraku** [comment, marked as answer]
Hey, the main reason I discourage from storing entities in your own container is because you would then be responsible to keep it up to date with the entities state (were they disposed? did their composition changed? ...) but if you know what you are doing it is completely valid to do so.
There is an existing container built in that could help you to achieve a spacial grid, the [`EntityMultiMap<>`](https://github.com/Doraku/DefaultEcs/blob/master/documentation/api/EntityMultiMap_TKey_.md). I actually use just that in the [boids sample](https://github.com/Doraku/DefaultEcs/blob/master/source/Sample/DefaultBoids/System/SetBehaviorSystem.cs#L31). I use a [`GridId`](https://github.com/Doraku/DefaultEcs/blob/master/source/Sample/DefaultBoids/Component/GridId.cs) component that is recalculated when the entities [move](https://github.com/Doraku/DefaultEcs/blob/master/source/Sample/DefaultBoids/System/MoveSystem.cs#L25) and when the multimap is notified of those changes it move entities around in what is basically a `Dictionary<GridId, Entity[]>` so I can access neighbors in O(1) (and as it is a native type of DefaultEcs it keeps track of any changes on the entities already).
Obviously this sample is really simple in 2d but the idea would be the same in 3d if needed.

**Comment by mc-lep** [comment]
Hi genaray & Doraku,

Thanks a lot, your solutions seems to fit what I want to do.
